doc/developer/design/20231127_decoupled_isolated_coordinator.md

```diff
@@ -0,0 +1,487 @@
+
+This section can be seen as an extension of
+[architecture-db.md](./../platform/architecture-db.md) but we drill down into
+the innards of ADAPTER and slightly change how we drive around COMPUTE AND
+STORAGE. At least we deviate from how they're driven around in the current
+physical design of Coordinator.
+
+> [!NOTE]
+> Please keep in mind that this section is about the _logical_ architecture. It
+> might seem verbose or over/under specified but is not necessarily indicative
+> of a performant implementation of the design.
+
+We introduce (or make explicit) two new components:
+
+- CATALOG maintains the persistent state of an environment and allows changing
+  and reading thereof. It is explicitly a pTVC that carries incremental changes
+  to the catalog!
+- TIMESTAMP ORACLE is a linearizable object that maintains read/write
+  timestamps for timelines.
+
+And we also change how some of the existing components work.
+
+### TIMESTAMP ORACLE
+
+The TIMESTAMP ORACLE can be considered a service that can be called from any
+component that needs it, and it provides linearizable timestamps and/or records
+writes as applied.
+
+- `ReadTs(timeline) -> Timestamp`: returns the current read timestamp.
+
+  This timestamp will be greater or equal to all prior values of
+  `ApplyWrite()`, and strictly less than all subsequent values of `WriteTs()`.
+
+- `WriteTs(timeline) -> Timestamp`: allocates and returns monotonically
+  increasing write timestamps.
+
+  This timestamp will be strictly greater than all prior values of `ReadTs()`
+  and `WriteTs`.
+
+- `ApplyWrite(timeline, timestamp)`: marks a write at time `timestamp` as
+  completed.
+
+  All subsequent values of `ReadTs()` will be greater or equal to the given
+  write `timestamp`.
+
+### CATALOG
+
+CATALOG can also be considered a shared service whose methods can be called
+from any component that needs it. And multiple actors can append data to it.
+
+Changes to the catalog are versioned in the `"catalog"` timeline, which is
+tracked separately from other timelines in TIMESTAMP ORACLE. Any successful
+append of updates to the catalog implicitly does an `ApplyWrite("catalog",
+new_upper)` on TIMESTAMP ORACLE.
+
+- `CompareAndAppend(expected_upper, new_upper, updates) -> Result<(),
+  AppendError>`: appends the given updates to the catalog _iff_ the expected
+  upper matches its current upper.
+
+- `SubscribeAt(timestamp) -> Subscribe`: returns a snapshot of the catalog at
+  `timestamp`, followed by an ongoing stream of subsequent updates.
+
+- `SnapshotAt(timestamp)`: returns an in-memory snapshot of the catalog at
+  `timestamp`.
+
+  A snapshot corresponds to our current in-memory representation of the catalog
+  that the Coordinator holds. This can be (and will be, in practice!)
+  implemented using `SubscribeAt` to incrementally keep an in-memory
+  representation of the catalog up to date with changes. However, it is
+  logically a different operation.
```

Comment:
I'm confused about the need for `SnapshotAt`! It seems to be a strict subset of `SubscribeAt`, because you could instead call `SubscribeAt`, wait for the initial snapshot to complete, and then terminate the subscription.

Reply:
Yep! I mention right here that it would be implemented in terms of `SubscribeAt`. I chose to keep it as a separate command (along with a description) here because I refer to "taking a catalog snapshot" below, where I describe how user queries are processed, plus I thought it might be a nice concept to keep in mind in general.

But I'm happy to remove it here if you think it doesn't add anything.